Added opcodes dispatch structure

Thordreck opened the pull request:
I have included a first approach to parse/dispath opcodes. Opcodes are split up in groups based on their first 4 bits. The actual code to execute on each opcode is implemented as specializations of a templated member function of the Interpreter class. This function is left undefined for the general value <T> case on purpose (you would get a linking error if you try to use the function with a value for which no explicit specialization exists).
The other branches (display, keypad, registers, etc) should be pulled into develop as well to be able to continue working on this.

Thordreck committed:
Added KeyMap class and tests.

Thordreck committed:
Added external friend parse function

Thordreck committed:
Merge branch 'develop' into opcodes

Thordreck committed:
Added empty specialization for each opcode type

blasco commented on src/Interpreter.cpp:
```diff
@@ -23,8 +24,18 @@ namespace chip8
-        //TODO: tick timers
+
+        //TODO: when should we stop?
+        while(true)
```
I think we should wrap this over a 'update' function and let another entity control the interpreter. This way we can stop it, change its speed etc.

blasco replied:
@Thordreck

blasco commented on src/details/instructions.cpp:
```diff
@@ -0,0 +1,44 @@
+        {
+            [] (Interpreter& _interpreter, const OpBytes& _op_bytes) { _interpreter.ExecuteInstruction<0x0>(_op_bytes); },
+            [] (Interpreter& _interpreter, const OpBytes& _op_bytes) { _interpreter.ExecuteInstruction<0x1>(_op_bytes); },
+            [] (Interpreter& _interpreter, const OpBytes& _op_bytes) { _interpreter.ExecuteInstruction<0x2>(_op_bytes); },
```
I wonder if we could get any use of using these templates, from my point of view here they are equivalent to a name, right? 
ExecuteInstruction<0x7> == ExecuteInstruction_0x7

I think we need to do some more parsing before dispatching the functions. For example:
these opcodes. 

8XY4 Add
8XY5 Subtract

With the current implementation we cannot have:
ExecuteInstruction_8XY4 and ExecuteInstruction_8XY5. 

We need to call  ExecuteInstruction<0x8> and the instruction needs to parse and see if it should do 8XY4 or 8XY5.
So the instruction needs to do a little bit of parsing work, that we might better move to the parser.

On the other hand, I don't know if there is any compilation time penalty for using ExecuteInstruction<0x8> rather than ExecuteInstruction_0x8. I don't mean for our case, but as a general practice. 

I still have to learn a lot about templates, but in case we want to keep them here, is it possible to do something like ExecuteInstruction<"8XY4"> ? To have some kind of tag to refer to them?

blasco replied:
@Thordreck

blasco committed:
Added some not working opcodes

blasco committed:
moved parse function from details to class

blasco committed:
Tested some OpCodes

blasco committed:
Merge branch 'master' into opcodes

blasco committed:
deleted unused files

blasco committed:
Wip

blasco committed:
added test file

Thordreck committed:
Moved opcodes and opbytes definition to a separate file. Added custom exception and simple strcat utility.

blasco committed:
wip

blasco committed:
wip

blasco committed:
Merge branch 'develop' into opcodes

blasco committed:
not working

blasco committed:
Added new opcodes with testcases

TestOpCode_0NNN
TestOpCode_00E0# Please enter the commit message for your changes. Lines starting
TestOpCode_00EE# with '#' will be ignored, and an empty message aborts the commit.
TestOpCode_1NNN#
TestOpCode_2NNN# On branch opcodes
TestOpCode_DXYN# Your branch is ahead of 'origin/opcodes' by 8 commits.
TestOpCode_ANNN#   (use "git push" to publish your local commits)

Thordreck committed:
Added opcode 3XNN (and tests).

Thordreck committed:
Added opcode 4XNN (and tests).

Thordreck committed:
Added opcode 5XY0 (and tests).

Thordreck committed:
Added opcode 6XNN (and tests).

Thordreck committed:
Added opcode 7XNN (and tests).

Thordreck committed:
Added opcode 8XY0 (and tests).

Thordreck committed:
Added opcode 8XY1 (and tests).

Thordreck committed:
Added opcode 8XY2 (and tests).

Thordreck committed:
Added opcode 8XY3 (and tests).

Thordreck committed:
Added opcode 8XY4 (and tests).

Thordreck committed:
Added opcode 8XY5 (and tests).

Thordreck committed:
Corrected carry logic on opcode 8XY5

Thordreck committed:
Added opcode 8XY6 (and tests).

Thordreck committed:
Added opcode 8XY7 (and tests).

Thordreck committed:
Added opcode 8XYE (and tests).

Thordreck committed:
Added opcode 9XY0 (and tests).

Thordreck committed:
Added opcode BNNN (and tests).

Thordreck committed:
Added opcode CXNN (and tests).

Thordreck committed:
Added opcode FX07 (and tests).

Thordreck committed:
Added opcode FX15 (and tests).

Thordreck committed:
Added opcode FX18 (and tests).

Thordreck committed:
Added opcode FX1E (and tests).

Thordreck committed:
Added opcode EX9E (and tests).

Thordreck committed:
Added opcode EXA1 (and tests).

Thordreck committed:
Added opcode FX0A.

Thordreck committed:
Added speaker in Interpreter. Added speaker callbacks to sound timer.

Thordreck committed:
main.cpp: load and start rom from input args.

blasco committed:
Added FX OpCodes

blasco committed:
Roms running, but sprites concentrated in one region

blasco committed:
Sprites rendering properly, weird behaviour after collision in Space Invaders

Thordreck committed:
Refactored timing control. Added logger facility. Added Frequency class. Improved strcat util.

Thordreck committed:
Added flag to notify when the display should be refreshed (this reduce CPU usage a bit).
Added unnamed parameters in opcodes that do not use their op_bytes to
suppress warnings.

blasco committed:
Fixed OpCodes bug. Shift left an right were wrong in the specification

blasco committed:
Added Chip8 Interpreter image

blasco committed:
Extended title generation analysis
